collect_app/src/androidTest/java/org/odk/collect/android/benchmark/EntitiesBenchmarkTest.kt

```diff
@@ -0,0 +1,137 @@
+package org.odk.collect.android.benchmark
+
+import android.app.Application
+import androidx.test.core.app.ApplicationProvider
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import org.hamcrest.MatcherAssert.assertThat
+import org.hamcrest.Matchers.blankOrNullString
+import org.hamcrest.Matchers.lessThan
+import org.hamcrest.Matchers.not
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.RuleChain
+import org.junit.runner.RunWith
+import org.odk.collect.android.support.TestDependencies
+import org.odk.collect.android.support.pages.MainMenuPage
+import org.odk.collect.android.support.pages.Page
+import org.odk.collect.android.support.rules.CollectTestRule
+import org.odk.collect.android.support.rules.TestRuleChain.chain
+import org.odk.collect.android.test.BuildConfig.ENTITIES_FILTER_TEST_PROJECT_URL
+import org.odk.collect.strings.R
+
+/**
+ * Benchmarks the performance of entity follow up forms. [PROJECT_URL] should be set to a project
+ * that contains the "100k Entities Filter" form.
+ *
+ * Devices that currently pass:
+ * - Pixel 4a
+ * - Fairphone 3
+ *
+ */
+
+@RunWith(AndroidJUnit4::class)
+class EntitiesBenchmarkTest {
+
+    private val rule = CollectTestRule(useDemoProject = false)
+
+    @get:Rule
+    var chain: RuleChain = chain(TestDependencies(true)).around(rule)
+
+    @Test
+    fun run() {
+        assertThat(
+            "Need to set ENTITIES_FILTER_TEST_PROJECT_URL before running!",
+            ENTITIES_FILTER_TEST_PROJECT_URL,
+            not(blankOrNullString())
+        )
+        clearAndroidCache()
+
+        val stopwatch = Stopwatch()
+
+        rule.startAtFirstLaunch()
+            .clickManuallyEnterProjectDetails()
+            .inputUrl(ENTITIES_FILTER_TEST_PROJECT_URL)
+            .addProject()
+
+            // Populate http cache and clear out form/entities
+            .clickGetBlankForm()
+            .clickGetSelected()
+            .clickOK(MainMenuPage())
+            .openProjectSettingsDialog()
+            .clickSettings()
+            .clickProjectManagement()
+            .clickOnResetProject()
+            .clickOnString(R.string.reset_blank_forms)
+            .clickOnString(R.string.reset_saved_forms)
+            .clickOnString(R.string.reset_settings_button_reset)
+            .clickOKOnDialog(MainMenuPage())
+
+            .clickGetBlankForm()
+            .benchmark("Downloading form with http cache", stopwatch) {
+                it.clickGetSelected()
+            }
+
+            .clickOK(MainMenuPage())
+            .clickGetBlankForm()
+            .benchmark("Updating form with http cache", stopwatch) {
```

Comment:
If it's the same form without any updates it's not really updating, right? Downloading will be skipped so what takes time here is comparing the differences. Am I right?

Reply:
Yeah, it should really be "Checking for updates when there are none" or something. I'll rename.